python_fundamentals-master/07_classes_objects_methods/07_02_shapes.py

```diff
@@ -9,27 +9,25 @@
 '''
 
 class rectangle:
+    square = True                   #class variable
     def __init__(self, length, width):
-        self.length = length
+        self.length = length        #attributes (variables that are only defined in init)
         self.width = width
-    def area(self):
         self.area = self.length * self.width
-        return print(f"the area of the rectangle is {self.area}")
-    def perimeter(self):
-        self.perimeter()  = 2  * (self.width + self.length)
-        return print(f"the perimeter is the rectangle is {self.perimeter}")
+        self.perimeter = 2 * (self.width + self.length)
+    def __str__(self):
+        return f"the area of the rectangle is {self.area} and the perimeter is the rectangle is {self.perimeter}"
 
 rec_1 = rectangle(5, 2)
-print(rec_1.area())
-print(rec_1.perimeter())
-
+print(rec_1)
 
 class circle:
-    def __init__(self, radius):
+    def __init__(self, radius, area, circumference):
         self.radius = radius
-    def area(self, length, width):
-        self.area = length * width
-        return print(f"the area of the circle is {self.area}")
-    def circumference(self, r):
-        self.circumference() = 2 * 3,1415 * self.radius
-        return print(f"the circumference of the circle is {self.circumference()}")
\ No newline at end of file
+        self.area = (3.1415 * self.radius ** 2)
+        self.circumference = 2 * 3.1415 * self.radius
```

Comment:
So, you don't need to pass `area, circumference` into the `init` constructor.  You can just create them in `__init__` without actually passing them when the class is created.

You should think about what gets passed as arguments to `__init__` as "what information does the user have to specify?
They do not need to specify the area or circumference since this can be solely calculated from the radius.
Does this make sense?

Basically you can do
```
def __init__(self, radius ):
        self.radius = radius
        self.area = (3.1415 * self.radius ** 2)
        self.circumference = 2 * 3.1415 * self.radius
```

And it will work the same.

Reply:
It does make sense, and I was struggling with this, because it didn't make sense to give it as an argument and it would calculate afterwards. For some reason I had errors before therefore I thought I had to specify all attributes before using them later on, but that is not needed. I changed it now and there are no errors anymore :) Good to know!